mesonbuild/modules/pkgconfig.py

```diff
@@ -435,17 +437,32 @@ def parse_variable_list(stringlist):
         variables = parse_variable_list(mesonlib.stringlistify(kwargs.get('variables', [])))
 
         pcfile = filebase + '.pc'
-        pkgroot = kwargs.get('install_dir', default_install_dir)
-        if pkgroot is None:
-            if mesonlib.is_freebsd():
-                pkgroot = os.path.join(state.environment.coredata.get_builtin_option('prefix'), 'libdata', 'pkgconfig')
-            else:
-                pkgroot = os.path.join(state.environment.coredata.get_builtin_option('libdir'), 'pkgconfig')
-        if not isinstance(pkgroot, str):
+        install_data = kwargs.get('install_data', False)
+        if install_data:
+            user_option = 'sharepkgconfigdir'
+        else:
+            user_option = 'libpkgconfigdir'
+        user_install_dir = state.environment.coredata.get_builtin_option(user_option)
+
+        install_dir = kwargs.get('install_dir', default_install_dir)
+        if install_dir is None:
+            install_dir = user_install_dir
+        elif install_dir != user_install_dir:
+            mlog.warning('''pkg-config file {} will be installed in {},
+but builtin option {} is set to {}.
+Since {} has a platform-dependen default value, it is usually
+a bad idea for a project to ignore the builtin option. To fix the issue for your
+project, you can either remove install_dir from both the pkgconfig.generate call
+and the library, or set install_dir to get_option('{}') to override
+the default value which may comes from the library.
+
+Please see https://mesonbuild.com/Pkgconfig-module.html#pkggenerate for more details.
```

Comment:
I don't think you can warn here, because there are legitimate cases where you want to put your pc file elsewhere. For example, GStreamer install pc files for plugins into `<libdir>/gstreamer-1.0/pkgconfig` for the case you want to static link to installed plugins. In that case user is responsible for setting PKG_CONFIG_PATH.